feat: add ica v3 docs (#76)
hyperlane-xyz/v3-docs: docs/reference/applications/ica.mdx

# Interchain Account Interface

Developers can use the **Interchain Account** interface to create and control an account on a remote chain from their local chain.

> Unlike general message passing, which requires recipients to implement a specific interface, Interchain Accounts (ICAs) allow developers to interact with _any remote_ contract.

Reviewer: ```suggestion
Unlike general message passing, which requires recipients to implement a specific interface, Interchain Accounts (ICAs) allow developers to interact with _any_ remote contract.
```

## Overview

### Diagram

```mermaid

flowchart TB
    subgraph Origin chain
    direction TB
      Sender
      A_O[InterchainAccountRouter]
      M_O[(Mailbox)]
    end

    subgraph Destination chain
    direction TB
      M_D[(Mailbox)]
      A_D[InterchainAccountRouter]
      SenderAccount
      Recipient
    end

    Sender -- "callRemote(destination, recipient, call)" --> A_O
    A_O -- "dispatch(destination, router, \n[sender, recipient, call])" --> M_O
    M_O -. "relay" .- M_D
    M_D -- "handle(origin, router, \n[sender, recipient, call])" --> A_D
    A_D == "interchainAccount(origin, sender)" ==> SenderAccount
    SenderAccount -- "call(data)" --> Recipient

    click A_O https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/solidity/contracts/middleware/InterchainAccountRouter.sol
    click A_D https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/solidity/contracts/middleware/InterchainAccountRouter.sol

    style Sender fill:#FF0099
    style SenderAccount fill:#FF0099
    style Recipient fill:#FF0099

```

> Interchain Accounts allow you to make a remote call from chain A to chain B using the router (`**InterchainAccountRouter**`). Here's how it works:

Reviewer: ```suggestion
Interchain Accounts allow you to make a remote call from **Chain A** to **Chain B** using the router (`InterchainAccountRouter`). Here's how it works:
```

> - We use [CREATE2](https://docs.openzeppelin.com/cli/2.8/deploying-with-create2) to compute the determinstic [OwnableMulticall](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/solidity/contracts/middleware/libs/OwnableMulticall.sol) contract address for you which serves as a proxy for your cross-chain calls. You can do it yourself too [here](#computing-addresses).

Reviewer: ```suggestion
- We use [CREATE2](https://docs.openzeppelin.com/cli/2.8/deploying-with-create2) to compute the determinstic [OwnableMulticall](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/solidity/contracts/middleware/libs/OwnableMulticall.sol) contract address for you, which serves as a proxy for your cross-chain calls. You can do it yourself too [here](#computing-addresses).
```

> - You can encode your call which includes the to address, call data, and the msg.value for each call, batched together in a array.

Reviewer: ```suggestion
- You can encode your call which includes the to address, call data, and the `msg.value` for each call, batched together in a array.
```

> - You send the encoded call to the chainA router which gets relayed to chainB router.

Reviewer: ```suggestion
- You send the encoded call to the **Chain A** router which gets relayed to the **Chain B** router.
```

> - After decoding the calls, the ChainB router checks if the computed address is already deployed or not. If not, we deploy the _OwnableMulticall_ contract.

Reviewer: ```suggestion
- After decoding the calls, the **Chain B** router checks if the computed address is already deployed or not. If not, we deploy the _OwnableMulticall_ contract.
```

> - The router then performs a multicall on the ICA address which in turn makes the desired arbitary call on chainB.

Reviewer: ```suggestion
- The router then performs a multicall on the ICA address, which in turn makes the desired arbitary call on **Chain B**.
```

The Interchain Account interface assigns every `(uint32 origin, address owner, address remoteRouter, address remoteISM)` tuple a unique ICA address. The sender owns that ICA on the destination chain, and can direct it to make arbitrary function calls via the `InterchainAccountRouter.callRemote()` endpoint.

> On many chains, you are able to use the defaults that are set by the owner of the router contract. See the [#overrides](#overrides) section to see how to make calls to any chain.

Reviewer: ```suggestion
For core chains supported by Hyperlane, you are able to use the defaults that are set by the owner of the router contract. See the [#overrides](#overrides) section to see how to make calls to _any_ chain.
```

---
description: Call any arbitrary existing contract on a remote chain
---

# InterchainAccount API

Developers can use the **InterchainAccount** API to create and control an account on a remote chain from their local chain.

Unlike the general message passing, which requires recipients to implement a specific interface, the InterchainAccount API allows developers to interact with any remote contract.

## Overview

### Diagram

```mermaid

flowchart TB
    subgraph Origin chain
    direction TB
      Sender
      A_O[InterchainAccountRouter]
      M_O[(Mailbox)]
    end

    subgraph Destination chain
    direction TB
      M_D[(Mailbox)]
      A_D[InterchainAccountRouter]
      SenderAccount
      Recipient
    end

    Sender -- "callRemote(destination, recipient, call)" --> A_O
    A_O -- "dispatch(destination, router, \n[sender, recipient, call])" --> M_O
    M_O -. "relay" .- M_D
    M_D -- "handle(origin, router, \n[sender, recipient, call])" --> A_D
    A_D == "interchainAccount(origin, sender)" ==> SenderAccount

>     SenderAccount -- "call(data)" --> Recipient

Author: fixed in the next pr

The InterchainAccount API assigns every `(uint32 origin, address owner, address remoteRouter, address remoteISM)` tuple a unique interchain account (ICA) address. The sender owns that ICA on the destination chain, and can direct it to make arbitrary function calls via the `InterchainAccountRouter.callRemote()` endpoint.

On many chains, you are able to use the defaults that are set by the owner of the router contract. See the [#overrides](#overrides) section to see how to make calls to any chain.

### Interface

```solidity
// SPDX-License-Identifier: MIT OR Apache-2.0
pragma solidity >=0.6.11;

import {CallLib} from "../contracts/libs/Call.sol";

interface IInterchainAccountRouter {
    function callRemote(
        uint32 _destinationDomain,
        CallLib.Call[] calldata calls
    ) external returns (bytes32);

    function getRemoteInterchainAccount(uint32 _destination, address _owner)
        external
        view
        returns (address);
}

```

:::tip

Want to use `InterchainAccountRouter` out of the box? Please refer to [addresses](../contract-addresses.mdx#interchainaccountrouter)

:::

> ## Example Usage

Author: won't match the rest of the docs, but I'll just link it as a tip.

> ## Example Usage

Author: actually added a section
